ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java

```diff
@@ -214,17 +214,61 @@ protected int getPartitions(final KafkaTopicClient kafkaTopicClient) {
   @Override
   void validateKeyPresent(final SourceName sinkName, final Projection projection) {
 
-    final boolean atLeastOneKey = joinKey.getAllViableKeys(schema).stream()
-        .anyMatch(projection::containsExpression);
+    if (joinKey.isForeignKey()) {
+      final DataSource leftInputTable = getLeftmostSourceNode().getDataSource();
+      final SourceName leftInputTableName = leftInputTable.getName();
+      final List<Column> leftInputTableKeys = leftInputTable.getSchema().key();
+      final List<Column> missingKeys =
+          leftInputTableKeys.stream().filter(
+              k -> !projection.containsExpression(
+                    new QualifiedColumnReferenceExp(leftInputTableName, k.name()))
+                  && !projection.containsExpression(new UnqualifiedColumnReferenceExp(
+                      ColumnNames.generatedJoinColumnAlias(leftInputTableName, k.name())
+                  ))
+          ).collect(Collectors.toList());
+
+      if (!missingKeys.isEmpty()) {
+        throwMissingKeyColumnException(missingKeys, leftInputTableName);
+      }
+    } else {
+      final boolean atLeastOneKey = joinKey.getAllViableKeys(schema).stream()
+          .anyMatch(projection::containsExpression);
 
-    if (!atLeastOneKey) {
-      final boolean synthetic = joinKey.isSynthetic();
-      final List<? extends Expression> viable = joinKey.getOriginalViableKeys(schema);
+      if (!atLeastOneKey) {
+        final boolean synthetic = joinKey.isSynthetic();
+        final List<? extends Expression> viable = joinKey.getOriginalViableKeys(schema);
 
-      throwKeysNotIncludedError(sinkName, "join expression", viable, false, synthetic);
+        throwKeysNotIncludedError(sinkName, "join expression", viable, false, synthetic);
+      }
     }
   }
 
+  private static void throwMissingKeyColumnException(
```

Comment:
nit: can we rename this to include FK joins in the name? Otherwise it's not clear what the difference between this method and `throwKeysNotIncludedError()` is since they do the same thing.

Reply:
Will rename if a follow up PR (assuming the build passes). If the build fails, will do it right away with any required fixed to make Jenkins pass.